Add PostGIS support

davidjayb opened the pull request:
In a similar vein to https://github.com/citusdata/django-multitenant/pull/68 I've added PostGIS support through a database backend and some additional models that reference GeoDjango models. GIS specific tests have been added and excluded from basic test runs as it requires the PostGIS extension to be installed.

gurkanindibay commented:
Hi @davidjayb 
Thanks for your valuable contribution
We recently added Codecov support to see our coverage. Could you rebase and get these developments as well to see the coverage effect of your contribution?
Thanks

davidjayb commented:
@gurkanindibay I've pulled in the changes but it doesn't look like the new workflows have been triggered. I will close and open up a new PR.

davidjayb commented:
I've reopened this PR. Not sure how to trigger the codecov workflows.

davidjayb commented:
@gurkanindibay I've added a pull request trigger to the workflow to get the tests to run. Running the workflow requires your approval:
https://github.com/citusdata/django-multitenant/actions/runs/4235453586

gurkanindibay commented:
@davidjayb Splendid effort. I like your work. Thanks for the contribution 👍. Actually, with your proposed solution we are getting two backends ; one for distributed usage and one for PostGIS usage, which should not be alternative to each other, in my opinion.
Is there a way to use both of them with just a the same usage for both distributed and PostGIS?
If you think it is not possible, still I want to work on unifying the solution. I can work on after prioritizing our tasks. 

Thanks

gurkanindibay commented:
BTW, tests are failing as well. Could you check them as well?

davidjayb commented on CONTRIBUTING.md:
````diff
@@ -48,3 +48,22 @@ Then in another shell run the tests:
 ```bash
````
Postgis is an extension of Postgresql that adds support for GIS data types and queries. This is not enabled by default. Citus uses base postgresql in their Docker images without support for Postgis:

https://docs.citusdata.com/en/latest/installation/single_node_docker.html

However Citus does support the Postgis extension:

https://docs.citusdata.com/en/latest/admin_guide/cluster_management.html#postgresql-extensions

But Postgis is not something everyone will want to enable. And not everyone wants to install GeoDjango, which requires additional OS packages such as GDAL (you can see the resulting failure that happens in the currently failing tests):
https://github.com/citusdata/django-multitenant/actions/runs/4235480906/jobs/7453429170

```
django.core.exceptions.ImproperlyConfigured: Could not find the GDAL library (tried "gdal", "GDAL", "gdal3.1.0", "gdal3.0.0", "gdal2.4.0", "gdal2.3.0", "gdal2.2.0", "gdal2.1.0", "gdal2.0.0"). Is GDAL installed? If it is, try setting GDAL_LIBRARY_PATH in your settings.
```

Therefore I think it makes sense to have a separate implementation for support both Postgres and Postgis.

davidjayb commented on django_multitenant/models.py:
```diff
@@ -20,3 +22,17 @@ class TenantModel(TenantModelMixin, models.Model):
         abstract = True
+
+
+class GisTenantManager(TenantManagerMixin, gis_models.Manager):
```
Yes, I think I can remove this. The `django.contrib.gis.db.models.Model` and `django.contrib.gis.db.models.Manager` end up being an alias for `django.db.models.Model` and `django.db.models.Manager`.

davidjayb commented on .github/workflows/django-multitenant-tests.yml:
```diff
@@ -4,6 +4,8 @@ on:
       - "**"
   workflow_dispatch:
+  pull_request:
+    types: [ opened, reopened, synchronize ]
```
Branches will only trigger in the main repository. For a PR like this one, which was created by a fork, we need to have a `pull_request` trigger to see the results in the PR. Is there another solution so that fork-based PRs will initiate a workflow?

davidjayb commented:
> @davidjayb Splendid effort. I like your work. Thanks for the contribution 👍. Actually, with your proposed solution we are getting two backends ; one for distributed usage and one for PostGIS usage, which should not be alternative to each other, in my opinion. Is there a way to use both of them with just a the same usage for both distributed and PostGIS? If you think it is not possible, still I want to work on unifying the solution. I can work on after prioritizing our tasks.
> 
> Thanks

@gurkanindibay I believe we need to have two separate backends. PostGIS is an extension that requires additional OS packages installed, like GDAL, that is not suitable for a Postgres database that does not require GIS related features.

My suggestion is that we keep the database backends separate. If the Citus team is willing to publish docker images with Postgis installed then we can use those for testing.

I'll work on refactoring with your feedback and fixing the test failures. Please let me know what you think.

gurkanindibay commented:
@davidjayb 
Thanks for all your efforts Nearly all tests have passed . 
There are some static code analysis failures
And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?

gurkanindibay commented:
> > @davidjayb Splendid effort. I like your work. Thanks for the contribution 👍. Actually, with your proposed solution we are getting two backends ; one for distributed usage and one for PostGIS usage, which should not be alternative to each other, in my opinion. Is there a way to use both of them with just a the same usage for both distributed and PostGIS? If you think it is not possible, still I want to work on unifying the solution. I can work on after prioritizing our tasks.
> > Thanks
> 
> @gurkanindibay I believe we need to have two separate backends. PostGIS is an extension that requires additional OS packages installed, like GDAL, that is not suitable for a Postgres database that does not require GIS related features.
> 
> My suggestion is that we keep the database backends separate. If the Citus team is willing to publish docker images with Postgis installed then we can use those for testing.
> 
> I'll work on refactoring with your feedback and fixing the test failures. Please let me know what you think.



> > @davidjayb Splendid effort. I like your work. Thanks for the contribution 👍. Actually, with your proposed solution we are getting two backends ; one for distributed usage and one for PostGIS usage, which should not be alternative to each other, in my opinion. Is there a way to use both of them with just a the same usage for both distributed and PostGIS? If you think it is not possible, still I want to work on unifying the solution. I can work on after prioritizing our tasks.
> > Thanks
> 
> @gurkanindibay I believe we need to have two separate backends. PostGIS is an extension that requires additional OS packages installed, like GDAL, that is not suitable for a Postgres database that does not require GIS related features.
> 
> My suggestion is that we keep the database backends separate. If the Citus team is willing to publish docker images with Postgis installed then we can use those for testing.
> 
> I'll work on refactoring with your feedback and fixing the test failures. Please let me know what you think.

@davidjayb we can first add the support then we can merge them under a single backend in another PR.In my opinion,  viving the users the chance of benefit from both PostGIS and Citus is a good option.

davidjayb commented:
> @davidjayb
> Thanks for all your efforts Nearly all tests have passed .
> There are some static code analysis failures
> And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?

I've added a step to run the GeoDjango tests after the Citus tests. This should cover the code I've added here.

gurkanindibay commented:
> > @davidjayb
> > Thanks for all your efforts Nearly all tests have passed .
> > There are some static code analysis failures
> > And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?
> 
> @gurkanindibay I've added a step to run the GeoDjango tests after the Citus tests. This should cover the code I've added here.



> > @davidjayb
> > Thanks for all your efforts Nearly all tests have passed .
> > There are some static code analysis failures
> > And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?
> 
> @gurkanindibay I've added a step to run the GeoDjango tests after the Citus tests. This should cover the code I've added here.

@davidjayb approved the CI execution and still have failures. Could you check these failures?

davidjayb commented:
> > > @davidjayb
> > > Thanks for all your efforts Nearly all tests have passed .
> > > There are some static code analysis failures
> > > And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?
> > 
> > 
> > @gurkanindibay I've added a step to run the GeoDjango tests after the Citus tests. This should cover the code I've added here.
> 
> > > @davidjayb
> > > Thanks for all your efforts Nearly all tests have passed .
> > > There are some static code analysis failures
> > > And there are some lines needs to be covered. Code coverage rate drops dramatically with this PR. Could you check the untested lines as well and add tests for them?
> > 
> > 
> > @gurkanindibay I've added a step to run the GeoDjango tests after the Citus tests. This should cover the code I've added here.
> 
> @davidjayb approved the CI execution and still have failures. Could you check these failures?

@gurkanindibay sorry I just realized that my fork runs GHA itself. The latest issue seems to be that the `django-base-image` job is not running. This is required to add the GDAL library to the base image:

https://github.com/davidjayb/django-multitenant/actions/runs/4316951694/jobs/7533400109#step:4:393

https://github.com/citusdata/django-multitenant/pull/150/files#diff-9b57afcb8b6128e7c6d80d6ed079cb7a11506b97a35ca9fbd0204ae619f205f0R21

gurkanindibay commented:
@davidjayb added the binary you requested and pushed into the dockerhub. Re-triggered the pipeline execution. There are some unit test errors. Could you check it again?

davidjayb commented:
@gurkanindibay alright hopefully this does it :-) I found out that Citus throws a DataError when a non-null constraint is violated whereas the base postgres/postgis database will throw an IntegrityError.

gurkanindibay commented:
> @gurkanindibay alright hopefully this does it :-) I found out that Citus throws a DataError when a non-null constraint is violated whereas the base postgres/postgis database will throw an IntegrityError.

Still errors exist could you check the tests again?

davidjayb commented:
@gurkanindibay I've been able to get all of these passing in my fork.

gurkanindibay commented:
> @gurkanindibay I've been able to get all of these passing in my fork:
> 
> https://github.com/davidjayb/django-multitenant/actions/runs/4357421331

thanks @davidjayb . All tests are passing in this PR as well. However, coverage is being dropped dramatically. Could you check the coverage reports in codecov https://github.com/citusdata/django-multitenant/pull/150/checks?check_run_id=11841895054

davidjayb commented:
@gurkanindibay this seems to be due to the code coverage only being used by the "Test GeoDjango MT" stage. There are several lines that are skipped over by the GIS tests as Citus is not being used.

Is there a way to merge the code coverage reports from the two test stages?

gurkanindibay commented:
@davidjayb thanks for all your efforts. Now I need to perform a detailed review.

gurkanindibay reviewed (commented):
@davidjayb Could you use our docker-compose file in the PR https://github.com/citusdata/django-multitenant/pull/164

davidjayb commented:
@gurkanindibay can you run the workflow again? The codecov failed because I forgot to remove a condition I introduced when running on a vanilla Postgis database (i.e. non-Citus).

gurkanindibay commented:
I was thinking the same as you. It is difficult for someone who wants to use both GIS and Citus together 
I will talk to our product manager tomorrow and in the meantime I will think about a solution. If you have any solution to work them together please commit it
Thanks for all your efforts

davidjayb commented:
@gurkanindibay in my mind having two separate backends is appropriate as the GIS backends are distinct in Django. Having the django-multitenant backends leverage mixins allow for the functionality to be common between the two.

gurkanindibay commented:
I will make one more round of review tomorrow

gurkanindibay commented:
@davidjayb I'm working on post-release issues and the blogpost about the new release. I will continue on reviewing after these tasks

gurkanindibay commented:
@davidjayb sorry for the delay of the review. I'm trying to handle post release tasks and prioritising bugs. I will return this task ASAP

davidjayb commented:
@gurkanindibay no problem. I'll wait to update the PR until the review is completed.

gurkanindibay commented on django_multitenant/backends/postgresql/base.py:
```diff
@@ -1,144 +1,21 @@
+class DatabaseSchemaEditor(
+    mixins.DatabaseSchemaEditorMixin, PostgresqlDatabaseSchemaEditor
+):
+    pass
```
We should keep the base class for regular models. They are required for tenancy of regular models

davidjayb replied:
I don't understand this feedback. This base class is still here. The mixin is used to extract common functionality that is used by both Postgres and Postgis.

gurkanindibay commented on django_multitenant/mixins.py:
```diff
@@ -217,3 +217,129 @@ def tenant_value(self):
         return get_object_tenant(self)
+
+
+class DatabaseSchemaEditorMixin:
```
Why did you need to move the classes in the base to here?

davidjayb replied:
See https://github.com/citusdata/django-multitenant/pull/150/#discussion_r1149440196 this is for reuse as these mixins provide the same functionality overrides for the Postgres and Postgis databases.

gurkanindibay commented on django_multitenant/tests/gis_models.py:
```diff
@@ -0,0 +1,22 @@
+import uuid
```
We need more tests to test all the cases
I think there should be a test for PostGIS models which will be correspond to Regular models. 
We are covering lots of cases  there

gurkanindibay commented on .github/workflows/django-multitenant-tests.yml:
```diff
@@ -64,6 +66,18 @@ jobs:
+        -v $(pwd):/build \
+        --network=host \
+        -e PYTHON_VERSION -e DJANGO_VERSION \
+        -e CODE_COV_APPEND="--cov-append" \
```
Since PostGIS and Citus have seperate tests and executions, we will need seperate coverage reports
Codecov has flags feature to handle this case
https://docs.codecov.com/docs/flags
Could you set the flag to see the PostGIS coverage seperately

gurkanindibay reviewed (commented):
@davidjayb I performed a round of review in detail. I added some comments and try to make comments as concise as possible. 
Thanks for your valuable contributions

davidjayb commented:
@gurkanindibay thank you for the review. Unfortunately I don't have the time to address the feedback now. I'll try to get back to this when I can.

davidjayb commented:
@gurkanindibay hey there, what is the status of the PR review? All checks have passed and there seems to be adequate coverage in the GIS tests.

davidjayb commented:
This PR has been sitting stale for quite some time so I have decided to close it. It doesn't seem like the team wants to take the time to get this merged in so hopefully you can take the time to add first-class support for PostGIS in the future.

gurkanindibay commented:
Hey @davidjayb 
Sorry for the long delay. I have other assignments and I couldn't deal with this PR until now.
You have put a great effort and PR is nearly OK to merge 
I have an available time slot this month. If you're available, we can add this PR and create a major release with this feature.
Open points in my mind
1. Can we use the PostGIS backend with models extended from TenantModel? Projects using django-multitenant with PostGIS backend may have addtional classes which does not include GIS attributes. What will be the behaviour in this case?
2. Now we have a detailed documentation https://django-multitenant.readthedocs.io/en/latest/. We need a special section here that defines project usage with PostGIS. After handling item 1, we need a clear documentation for users to easily use the GIS feature

I hope you can reserve time for this PR. Thanks for all of your contributions @davidjayb 🥇
